src/main/java/org/hisp/dhis/rules/functions/RuleFunctionExtractDataMatrixValue.java

```diff
@@ -57,7 +57,8 @@ public Object evaluate( ExpressionParser.ExprContext ctx, CommonExpressionVisito
     public Object getDescription( ExpressionParser.ExprContext ctx, CommonExpressionVisitor visitor )
     {
 
-        extractDataMatrixValue( visitor.castStringVisit( ctx.expr( 1 ) ), visitor.castStringVisit( ctx.expr( 0 ) ) );
+        visitor.castStringVisit( ctx.expr( 1 ) );
```

Comment:
@enricocolasante Good point! That makes sense to me.

Reply:
Yes, that perfectly makes sense. getDescription only needs syntactic expression validation and for that, we don't need to actually execute the method functionality. `extractDataMatrixValue` will actually trigger the function. For syntactic validation, we only need to traverse or visit the expression.

Reply:
I understood @enricocolasante's comment to say that if the `extractDataMatrixValue` method did some additional checking of the expression syntax, we should call it on `getDescription` as well as on `evaluate`. We want to find out as soon as possible if there is anything wrong, and not wait until run time. This made sense to me and I said so, and re-requested my own review.

On reading @zubaira's comment and on looking more closely at the code, here is what I see: In the parser, `d2:extractDataMatrixValue` takes two `expr` arguments which could come from any number of places: constant text, data values, results of an if statement, etc. As such, we cannot know the values until `evaluate` is called. This being the case, it makes no sense to try and further validate the arguments during `getDescription` because we cannot know what they will be. In this case, @zubaira's code and his comment make perfect sense.

However on reading [DHIS2-12364](https://jira.dhis2.org/browse/DHIS2-12364) I have a question for @Bekkalizer, if he is available to answer it:

@Bekkalizer, in the ticket, you give this example of an expression using this function:

`d2:extractDataMatrixValue('gtin',A{GS1 Value})`

Would you like the parser to restrict the first argument (the GS1 key) so that it must be a string constant, or would you like it to be any general expression that evaluates to a string (for example, a string constant, a program attribute, a program indicator, the result of an `if()` function, etc.) Some pros and cons:

**string constant**: If we restrict the GS1 key to a string constant, then during `getDescription` we could validate that the constant is a valid GS1 key that we support. However, @zubaira would have to change the parser to restrict the first argument to a string constant, and change the rule engine to evaluate the first argument during `getDescription`. (The second argument, the GS1 value, would need to not be evaluated in `getDescription`.)

**general expression**: If the GS1 key is a general expression, then it could be a string constant, a program attribute, etc., etc. In this case it would not make sense to try and validate it at `getDescription` time. At present, the code assumes this, both in the parser and in the rule engine with this PR, so no additional changes would be needed for this.

Reply:
@jimgrace I can answer the question, I could not assign the ticket to myself so Markus did in order to move it through the workflow. :)

The GS1 key is indeed a string constant that can have a (very long) set number of values. but I set this first argument in the parser as an expression to allow populating this argument from another variable if that's what the implementer wants. Like for example, having and option set to select what GS1 element they want to extract.

Reply:
Sounds good. Thanks @Balcan!